File: data_structures/stacks_and_queues/stack_with_max.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cassert>
#include <algorithm>

using std::cin;
using std::string;
using std::vector;
using std::cout;
using std::max_element;

class StackWithMax {
    vector<int> stack;
    vector<int> auxiliary_stack;

  public:

    void Push(int value) {
        stack.push_back(value);
        int a_stack_size = auxiliary_stack.size();

        if (a_stack_size > 0) {
            int last_symbol = auxiliary_stack[a_stack_size - 1];

            if (value > last_symbol) {
              auxiliary_stack.push_back(value);
            } else {
              auxiliary_stack.push_back(last_symbol);
            }
          } else {
            auxiliary_stack.push_back(value);
          }
    }

    void Pop() {
        assert(stack.size());
        stack.pop_back();
        auxiliary_stack.pop_back();
    }

    int Max() const {
        int stack_size = stack.size();
        assert(stack_size > 0);
        return auxiliary_stack[stack_size - 1];
    }
};

int main() {
    int num_queries = 0;
    cin >> num_queries;

    string query;
    string value;

    StackWithMax stack;

    for (int i = 0; i < num_queries; ++i) {
        cin >> query;
        if (query == "push") {
            cin >> value;
            stack.Push(std::stoi(value));
        }
        else if (query == "pop") {
            stack.Pop();
        }
        else if (query == "max") {
            cout << stack.Max() << "\n";
        }
        else {
            assert(0);
        }
    }
    return 0;
}

File: data_structures/stacks_and_queues/stack_with_max.py
# python3
"""
Implement a stack supporting the operations Push(),Pop(), andMax().

INPUT: The first line of the input contains the number q of queries.
Each of the following q lines specifies a query of one of the following formats: push v, pop, or max

OUTPUT: For each max query, output (on a separate line) the maximum value of the stack
"""
import sys


class StackWithMax():
    def __init__(self):
        self.stack = []
        self.auxiliary_stack = []

    def push(self, a):
        self.stack.append(a)

        if len(self.auxiliary_stack) == 0:
            self.auxiliary_stack.append(a)
        else:
            last_symbol = self.auxiliary_stack[-1]
            if a > last_symbol:
                self.auxiliary_stack.append(a)
            else:
                self.auxiliary_stack.append(last_symbol)

    def pop(self):
        assert (len(self.stack)) > 0
        self.stack.pop()
        self.auxiliary_stack.pop()

    def max(self):
        assert (len(self.stack)) > 0
        return self.auxiliary_stack[-1]


if __name__ == '__main__':
    stack = StackWithMax()
    _num_queries = sys.stdin.readline()

    for line in sys.stdin.readlines():
        command, *value = line.split()

        if command == "push":
            stack.push(int(*value))

        elif command == "pop":
            stack.pop()

        elif command == "max":
            print(stack.max())
